cpp/src_prims/timeSeries/arima_helpers.cuh

```diff
@@ -52,6 +52,117 @@ HDI DataT _param_to_poly(const DataT* param, int lags, int idx)
+ * @param[in] size0 Size of src0
+ * @param[in] src1  Data comes from here if the index is positive
+ * @param[in] idx   Index, relative to the start of the second array src1
+ * @return Data read from src0 or src1 according to the index
+ */
+template <typename DataT>
+DI DataT _select_read(const DataT* src0, int size0, const DataT* src1, int idx)
+{
+  return idx < 0 ? src0[size0 + idx] : src1[idx];
+}
+
+/**
+ * @brief Prepare future data with a simple or seasonal difference
+ *
+ * @param[in]  in_past  Input (past). Shape (n_past, batch_size) (device)
+ * @param[in]  in_fut   Input (future). Shape (n_fut, batch_size) (device)
+ * @param[out] out      Output. Shape (n_fut, batch_size) (device)
+ * @param[in]  n_past   Number of past observations per series
+ * @param[in]  n_fut    Number of future observations per series
+ * @param[in]  period   Differencing period (1 or s)
+ * @param[in]  stream   CUDA stream
+ */
+template <typename T>
+__global__ void _future_diff_kernel(
+  const T* in_past, const T* in_fut, T* out, int n_past, int n_fut, int period = 1)
+{
+  const T* b_in_past = in_past + n_past * blockIdx.x;
+  const T* b_in_fut  = in_fut + n_fut * blockIdx.x;
+  T* b_out           = out + n_fut * blockIdx.x;
+
+  for (int i = threadIdx.x; i < n_fut; i += blockDim.x) {
+    b_out[i] = b_in_fut[i] - _select_read(b_in_past, n_past, b_in_fut, i - period);
+  }
+}
+
+/**
+ * @brief Prepare future data with two simple and/or seasonal differences
+ *
+ * @param[in]  in_past  Input (past). Shape (n_past, batch_size) (device)
+ * @param[in]  in_fut   Input (future). Shape (n_fut, batch_size) (device)
+ * @param[out] out      Output. Shape (n_fut, batch_size) (device)
+ * @param[in]  n_past   Number of past observations per series
+ * @param[in]  n_fut    Number of future observations per series
+ * @param[in]  period1  First differencing period (1 or s)
+ * @param[in]  period2  Second differencing period (1 or s)
+ * @param[in]  stream   CUDA stream
+ */
+template <typename T>
+__global__ void _future_second_diff_kernel(const T* in_past,
+                                           const T* in_fut,
+                                           T* out,
+                                           int n_past,
+                                           int n_fut,
+                                           int period1 = 1,
+                                           int period2 = 1)
+{
+  const T* b_in_past = in_past + n_past * blockIdx.x;
+  const T* b_in_fut  = in_fut + n_fut * blockIdx.x;
+  T* b_out           = out + n_fut * blockIdx.x;
+
+  for (int i = threadIdx.x; i < n_fut; i += blockDim.x) {
+    b_out[i] = b_in_fut[i] - _select_read(b_in_past, n_past, b_in_fut, i - period1) -
+               _select_read(b_in_past, n_past, b_in_fut, i - period2) +
+               _select_read(b_in_past, n_past, b_in_fut, i - period1 - period2);
+  }
+}
+
+/**
+ * @brief Kernel to undifference the data with up to two levels of simple
```

Comment:
Either here, or at `finalize_forcast`, it would be good to extend the description of the operation by relating it to "I" in ARIMA.